src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -71,45 +71,101 @@ internal static class DotNetSdkLocationHelper
         }
 
         public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirectory, bool allowQueryAllRuntimes)
-        {            
-            foreach (var basePath in GetDotNetBasePaths(workingDirectory))
+        {
+            string? bestSdkPath;
+            string[] allAvailableSdks;
+            try
+            {
+                AddUnmanagedDllResolver();
+
+                bestSdkPath = GetSdkFromGlobalSettings(workingDirectory);
+                allAvailableSdks = GetAllAvailableSDKs().ToArray();
+            }
+            finally
+            {
+                RemoveUnmanagedDllResolver();
+            }
+
+            Dictionary<Version, VisualStudioInstance?> versionInstanceMap = new();
+            foreach (var basePath in allAvailableSdks)
             {
                 var dotnetSdk = GetInstance(basePath, allowQueryAllRuntimes);
                 if (dotnetSdk != null)
                 {
-                    yield return dotnetSdk;
+                    // We want to return the best SDK first
+                    if (dotnetSdk.VisualStudioRootPath == bestSdkPath)
+                    {
+                        // We will add a null entry to the map to ensure we do not add the same SDK from a different location.
+                        versionInstanceMap[dotnetSdk.Version] = null;
+                        yield return dotnetSdk;
+                    }
+
+                    // Only add an SDK once, even if it's installed in multiple locations.
+                    if (!versionInstanceMap.ContainsKey(dotnetSdk.Version))
+                    {
+                        versionInstanceMap.Add(dotnetSdk.Version, dotnetSdk);
+                    }
```

Comment:
Tiny nit:
I think we can skip to:
```suggestion
                    // Only add an SDK once, even if it's installed in multiple locations.
                    versionInstanceMap[dotnetSdk.Version] = dotnetSdk;
```

(One fewer dictionary access)

Reply:
So I am unsure if it is a benefit or not, by keeping the first SDK for each discovered version we maintain a preference for SDKs in the following install location ordering.

1. DOTNET_ROOT
2. Current process path if MSBuild.Locator is called from dotnet.exe
3. DOTNET_HOST_PATH
4. DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR
5. PATH

Reply:
I'm trying to figure out if it matters or not, and I haven't come up with an answer, so playing it safe is probably a good idea.

Reply:
`Dictionary<>.TryAdd()` would have that behavior right?

Reply:
Isn't this code used only on core?

Reply:
Good point 👍

Reply:
I had thought about TryAdd and talked myself out of it without realizing this code only builds under .NET.

Reply:
In your defense the repo has "MSBuild" in the name so thinking everything targets v.OldAndBusted is not a bad assumption!